### PR TITLE
Add Calendar Event creation and deletion dialogs

### PR DESCRIPTION
**Pull request by allenporter**
## Proposed change

Add Calendar Event dialogs for creation and deletion.  This is part of a change to support local calendar integration, but also is intended to work for others like google calendar, caldav, etc.

This is a quick overview of the relevant parts of core PRs to aid in reviewing this change:
- A calendar may support mutation with a supported feature
- The existing date range picker was modified to support day vs day + time picking
- Recurring events are handled by the integration, flattened to individual event records. However, core will pass up a string rrule that the frontend can interpret to allow recurrence event ui editing (e.g. `FREQ=DAILY;COUNT=3`)
- An event is identified by a `uid`. A recurring event also has a `recurrence_id` to specify an individual instance in the series
- Recurring events may be updated or deleted in a few ways following rfc5545 field names and rfc5546 patterns for communication between calendar systems:
  - Specify `uid` only to delete the entire series
  - Specify `uid` and `recurrence_id` to handle a single instance in the series
  - Specify `uid` and `recurrence_id` and a `recurrence_range` or `THISANDFUTURE` to delete future events.

Core PR:  https://github.com/home-assistant/core/pull/79601
Architecture discussion: https://github.com/home-assistant/architecture/discussions/797

Work in progress for this PR:
- [X] Create and delete APIs
- [X] Update the date range picker to support all day events w/o times
- [X] Initially RRULE editing as a string
- [X] Simplify confirmation dialog bog handling for confirmation
- [X] Fix start/end date editing for non-all day events
- [X] Fix bugs in all day event handling
- [X] Fix field displays for info panel when not creating
- [X] Subset of RRULE editing UI

Work for future PRs:
- [ ] More RRULE editing (e.g. recurring by day of week, day of month, etc). For now, only works on local calendar so its ok to support a subset.
- [ ] Add more rfc5545 fields

## Type of change

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:
  - https://github.com/home-assistant/home-assistant.io/pull/24429
  - https://github.com/home-assistant/developers.home-assistant/pull/1488
  - 
Add Event
Add All Day Repeated Event
View Repeated Event
View All Day Repeated Event

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by allenporter** [comment]
I've gotten pretty far along on this, but it still needs a bunch of polish (e.g. maybe dialog is too small, sizing isn't right, perhaps some better formatting). I thought it was worth getting feedback at this point before I get much further.

File: src/components/ha-date-range-picker.ts [hunk 1]
```diff
@@ -54,6 +59,8 @@ export class HaDateRangePicker extends LitElement {
     return html`
       <date-range-picker
         ?disabled=${this.disabled}
+        auto-apply=${this.autoApply}
+        time-picker=${this.timePicker}
```
Comment (bramkragten) [inline src/components/ha-date-range-picker.ts:63]:
```suggestion
        ?auto-apply=${this.autoApply}
        ?time-picker=${this.timePicker}
```

File: src/components/ha-date-range-picker.ts [hunk 2]
```diff
@@ -62,7 +69,9 @@ export class HaDateRangePicker extends LitElement {
         <div slot="input" class="date-range-inputs">
           <ha-svg-icon .path=${mdiCalendar}></ha-svg-icon>
           <ha-textfield
-            .value=${formatDateTime(this.startDate, this.hass.locale)}
+            .value=${this.timePicker === true
```
Comment (bramkragten) [inline src/components/ha-date-range-picker.ts:72]:
```suggestion
            .value=${this.timePicker
```

File: src/components/ha-date-range-picker.ts [hunk 3]
```diff
@@ -71,7 +80,9 @@ export class HaDateRangePicker extends LitElement {
             readonly
           ></ha-textfield>
           <ha-textfield
-            .value=${formatDateTime(this.endDate, this.hass.locale)}
+            .value=${this.timePicker === true
```
Comment (bramkragten) [inline src/components/ha-date-range-picker.ts:83]:
```suggestion
            .value=${this.timePicker
```

File: src/panels/calendar/dialog-calendar-event-detail.ts [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+                      id="all_day"
+                      .checked=${this._allDay}
+                      @change=${this._allDayToggleChanged}
+                    ></ha-switch>
+                  </ha-formfield>
+
+                  <ha-rrule
+                    .hass=${this.hass}
+                    .value=${this._data!.rrule}
+                    .mutable=${true}
```
Comment (bramkragten) [inline src/panels/calendar/dialog-calendar-event-detail.ts:184]:
```suggestion
                    mutable
```

File: src/panels/calendar/dialog-calendar-event-detail.ts [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+            : html`
+                <div class="field">
+                  <ha-svg-icon .path=${mdiCalendarClock}></ha-svg-icon>
+                  <div class="value">
+                    ${this._formatDateRange()}<br />
+                    ${this._data!.rrule !== undefined
+                      ? html` <ha-rrule
+                          .hass=${this.hass}
+                          .value=${this._data!.rrule}
+                          .mutable=${false}
```
Comment (bramkragten) [inline src/panels/calendar/dialog-calendar-event-detail.ts:222]: Maybe make this `disabled`?

Comment (bramkragten) [inline src/panels/calendar/dialog-calendar-event-detail.ts:222]:
```suggestion
```

File: src/panels/calendar/dialog-calendar-event-detail.ts [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+                        </ha-rrule>`
+                      : html``}
+                  </div>
+                </div>
+
+                <div class="attribute">
+                  <state-info
+                    .hass=${this.hass}
+                    .stateObj=${stateObj}
+                    .inDialog="true"
```
Comment (bramkragten) [inline src/panels/calendar/dialog-calendar-event-detail.ts:233]:
```suggestion
                    inDialog
```

**Comment by bramkragten** [comment]
Have not dived into this deep enough yet, but have you seen <https://fullcalendar.io/docs/rrule-plugin>? Not sure if it is helpful here

File: src/panels/calendar/confirm-event-dialog-box.ts [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+      >
+        <div>
+          <p>${this._params.text}</p>
+        </div>
+        <mwc-button @click=${this._dismiss}>
+          ${this.hass.localize("ui.dialogs.generic.cancel")}
+        </mwc-button>
+        <mwc-button
+          @click=${this._confirm}
+          ?dialogInitialFocus="true"
```
Comment (bramkragten) [inline src/panels/calendar/confirm-event-dialog-box.ts:49]:
```suggestion
          dialogInitialFocus
```

File: src/panels/calendar/confirm-event-dialog-box.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        ?escapeKeyAction="true"
+        @closed=${this._dialogClosed}
+        defaultAction="ignore"
+        hideActions="false"
+        .heading=${this._params.title}
+      >
+        <div>
+          <p>${this._params.text}</p>
+        </div>
+        <mwc-button @click=${this._dismiss}>
```
Comment (bramkragten) [inline src/panels/calendar/confirm-event-dialog-box.ts:44]:
```suggestion
        <mwc-button @click=${this._dismiss} slot="secondaryAction">
```

File: src/panels/calendar/confirm-event-dialog-box.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+          ${this.hass.localize("ui.dialogs.generic.cancel")}
+        </mwc-button>
+        <mwc-button
```
Comment (bramkragten) [inline src/panels/calendar/confirm-event-dialog-box.ts:47]:
```suggestion
        <mwc-button slot="primaryAction"
```

File: src/panels/calendar/confirm-event-dialog-box.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+          @click=${this._confirm}
+          ?dialogInitialFocus="true"
+          class="destructive"
+        >
+          ${this._params.confirmText}
+        </mwc-button>
+        ${this._params.confirmFutureText
+          ? html`
+              <mwc-button @click=${this._confirmFuture} class="destructive">
```
Comment (bramkragten) [inline src/panels/calendar/confirm-event-dialog-box.ts:56]:
```suggestion
              <mwc-button @click=${this._confirmFuture} class="destructive"  slot="primaryAction">
```

File: src/panels/calendar/dialog-calendar-event-detail.ts [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+                <mwc-button
+                  slot="secondaryAction"
+                  class="warning"
+                  @click=${this._deleteEvent}
+                  .disabled=${this._submitting}
+                >
+                  ${this.hass.localize("ui.components.calendar.event.delete")}
+                </mwc-button>
+              `}
+        </div>
```
Comment (bramkragten) [inline src/panels/calendar/dialog-calendar-event-detail.ts:246]: The buttons should not be inside the div

File: src/panels/calendar/ha-rrule.ts [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+  @state() private _freq?: RepeatFrequency;
+
+  @state() private _rrule?: Partial<Options>;
+
+  protected firstUpdated() {
+    if (this.value === undefined || this.value === "") {
+      this._freq = RepeatFrequency.NONE;
+      this._rrule = undefined;
+      this.requestUpdate();
```
Comment (bramkragten) [inline src/panels/calendar/ha-rrule.ts:69]: No need for `requestUpdate`

File: src/panels/calendar/ha-rrule.ts [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+  @property({ type: Boolean }) public mutable = true;
+
+  @property() public value?: string;
+
+  @state() private _freq?: RepeatFrequency;
+
+  @state() private _rrule?: Partial<Options>;
+
+  protected firstUpdated() {
```
Comment (bramkragten) [inline src/panels/calendar/ha-rrule.ts:65]: Don't use `firstUpdated` but use `willUpdate` and check if `value` was changed:
```ts
protected willUpdate(changedProps) {
  if (changedProps.has("value")) {
   ... calc `_freq` and `_rrule`
  }
}
```

File: src/panels/calendar/show-confirm-event-dialog-box.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+      dialogTag: "confirm-event-dialog-box",
+      dialogImport: loadGenericDialog,
+      dialogParams: {
+        ...dialogParams,
+        confirm: (thisAndFuture: RecurrenceRange) => {
+          resolve(thisAndFuture);
+        },
+        cancel: () => {
+          resolve(undefined);
+        },
```
Comment (bramkragten) [inline src/panels/calendar/show-confirm-event-dialog-box.ts:31]: Why hardcode this here?

**Comment by allenporter** [comment]
> Have not dived into this deep enough yet, but have you seen https://fullcalendar.io/docs/rrule-plugin? Not sure if it is helpful here

I have seen this. My impression is that this helps with rendering an event in the display with a recurrence rule. If we, for example, had core only know about the primary event we could get it to display all the instance with this plugin in the UI. However, we also want core to know about the expanded rrule so the local calendar integration does the recurrence rule evaluation when queried.

I think the main thing we want from a library would be UI for generating RRULE strings.  Here are some otions i've looked at and would love your input.

# React RRule Generator

https://fafruch.github.io/react-rrule-generator/ -- This seems like what we'd want?  However, that github project is not active which made me think we wouldn't want to depend on this https://github.com/fafruch/react-rrule-generator#react-rrule-generator . It has some controls then outputs the `RRULE` string to give to core. I basically was thinking we'd want to copy this.

# RRule js

https://jakubroztocil.github.io/rrule/ -- seems a little too low level to be useful.

File: src/panels/calendar/confirm-event-dialog-box.ts (hunk 9 above)
Comment (allenporter) [reply src/panels/calendar/confirm-event-dialog-box.ts]: After fixing this, it looks like this:

Screen Shot 2022-10-17 at 8 51 58 PM

Is that red outline OK?

(I also removed the `hideActions` above)

File: src/panels/calendar/dialog-calendar-event-detail.ts (hunk 5 above)
Comment (allenporter) [reply src/panels/calendar/dialog-calendar-event-detail.ts]: Reamed this field.

Comment (allenporter) [reply src/panels/calendar/dialog-calendar-event-detail.ts]: Renamed this field, and removed this.

File: src/panels/calendar/ha-rrule.ts (hunk 12 above)
Comment (allenporter) [reply src/panels/calendar/ha-rrule.ts:69]: Removed.

File: src/panels/calendar/dialog-calendar-event-detail.ts (hunk 11 above)
Comment (allenporter) [reply src/panels/calendar/dialog-calendar-event-detail.ts:224]: Moved.

File: src/panels/calendar/ha-rrule.ts (hunk 13 above)
Comment (allenporter) [reply src/panels/calendar/ha-rrule.ts]: OK, i think where i was having trouble was also handling initialization. I've set an initial value for `_freq` instead which seems to address that.

File: src/panels/calendar/show-confirm-event-dialog-box.ts (hunk 14 above)
Comment (allenporter) [reply src/panels/calendar/show-confirm-event-dialog-box.ts:31]: My goal is just to simplify the caller logic so this works:

```
const ....  = await showConfirmEventDialog(this, ...);
if (.....) {   // Cancel
    return
} else {
  try {
    await deleteCalendarEvent(
      ...  // delete this event, or delete this and future
    );
  } catch (err: any) {
...
```

The approach I took was copied from an existing dialog, but i don't like it. Can you give me a suggestion on another example to look at or a better structure?

File: src/panels/calendar/confirm-event-dialog-box.ts (hunk 9 above)
Comment (bramkragten) [reply src/panels/calendar/confirm-event-dialog-box.ts:46]: Yeah, the red outline is ok, would maybe only swap `all` and `only this` buttons

File: src/panels/calendar/show-confirm-event-dialog-box.ts (hunk 14 above)
Comment (bramkragten) [reply src/panels/calendar/show-confirm-event-dialog-box.ts:31]: You copied it from the confirm dialog I guess?

What you miss here is the
```ts
    const origCancel = dialogParams.cancel;
    const origConfirm = dialogParams.confirm;
```

and then all those in your overwrite of `confirm` and `cancel`

Comment (allenporter) [reply src/panels/calendar/show-confirm-event-dialog-box.ts]: Ah yes, i removed that since it was unused, but now see why thats confusing. Fixed.

**Comment by allenporter** [comment]
Thank you for reviewing all the basics of this... It may be soon time to start talking about how this thing looks?
- The form boxes may be a bit too close together, so it may need some styling 
- is this too large to fit in a dialog or should the event editor move to an entire panel?

I'll also likely plan to add more to this, like https://fafruch.github.io/react-rrule-generator/ (days of the week, day of month, end date or # of occurrences)

Any other high level guidance you have here would be appreciated and whether it makes sense to get right here vs follow up.

**Comment by allenporter** [comment]
@bramkragten I attempted to fix the z-index for the date-range picker but it ended up also making the date text  boxes extend over the dialog to the right side. perhaps you can give me some more guidance on the best approach for this as it seems like i may need to make style changes within the date range picker component.

Generally, maybe we can chat about revisiting next steps here, also happy to chat on discord.

**Comment by bramkragten** [comment]
I would say we build our own, having a react component again is not something I really want, and if it is not maintained anyway 🤷‍♂️ 

I'll help you build it

**Comment by allenporter** [comment]
I ended up getting a book on web components and building my own https://allenporter.github.io/recurrence-rule-editor/

I'll update this PR to integrate that, though it just needs a little more: translation support and some stylings, plus any other observations you have. Also happy to host elsewhere if you want it part of home assistant.

**Review by bramkragten — APPROVED** [review]

**Pull request merged**